packages/flutter_tools/lib/src/android/android_emulator.dart

```diff
@@ -117,41 +116,45 @@ class AndroidEmulators extends EmulatorDiscovery {
 
 class AndroidEmulator extends Emulator {
   AndroidEmulator(String id, {
-    Map<String, String> properties,
-    @required Logger logger,
-    @required AndroidSdk androidSdk,
-    @required ProcessManager processManager,
+    Map<String, String>? properties,
+    required Logger logger,
+    AndroidSdk? androidSdk,
+    required ProcessManager processManager,
   }) : _properties = properties,
        _logger = logger,
        _androidSdk = androidSdk,
        _processUtils = ProcessUtils(logger: logger, processManager: processManager),
        super(id, properties != null && properties.isNotEmpty);
 
-  final Map<String, String> _properties;
+  final Map<String, String>? _properties;
   final Logger _logger;
   final ProcessUtils _processUtils;
-  final AndroidSdk _androidSdk;
+  final AndroidSdk? _androidSdk;
 
   // Android Studio uses the ID with underscores replaced with spaces
   // for the name if displayname is not set so we do the same.
   @override
   String get name => _prop('avd.ini.displayname') ?? id.replaceAll('_', ' ').trim();
 
   @override
-  String get manufacturer => _prop('hw.device.manufacturer');
+  String? get manufacturer => _prop('hw.device.manufacturer');
 
   @override
   Category get category => Category.mobile;
 
   @override
   PlatformType get platformType => PlatformType.android;
 
-  String _prop(String name) => _properties != null ? _properties[name] : null;
+  String? _prop(String name) => _properties != null ? _properties![name] : null;
 
   @override
-  Future<void> launch({@visibleForTesting Duration startupDuration, bool coldBoot = false}) async {
+  Future<void> launch({@visibleForTesting Duration? startupDuration, bool coldBoot = false}) async {
+    final String? emulatorPath = _androidSdk?.emulatorPath;
+    if (emulatorPath == null) {
+      throwToolExit('Emulator is missing from the Android SDK');
```

Comment:
should this be a toolExit? could this be Android Studio moving things from under us?

Reply:
Swapped to a `throw Exception`.